File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int solve(TreeNode* root,int & ans){
        if(root==NULL) return 0;
        int right=solve(root->right,ans);
        int left=solve(root->left,ans);
        ans=max(ans,right+left);
        return 1+max(right,left);
    }
    int diameterOfBinaryTree(TreeNode* root) {
        int ans=0;
        int height=solve(root,ans);
        return ans;
    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans=0;
    public int solve(TreeNode root){
        if(root==null) return 0;
        int right=solve(root.right);
        int left=solve(root.left);
        ans=Math.max(ans,right+left);
        return 1+Math.max(right,left);
    }
    public int diameterOfBinaryTree(TreeNode root) {
        int height=solve(root);
        return ans;
    }
}